### ISSUE TITLE
EF7 Performance Regression when using large number of 'OR' Predicates in Expression

### ISSUE THREAD

**Issue by npnelson** [aspnet/EntityFramework#6254]
### Steps to reproduce
https://github.com/npnelson/ExpressionSample
<b> Note you need SQL LocalDb to run sample, I used SQL2016 localDb</b>

### The issue
I've boiled down my program to the bare minimum to reproduce the issue.  My sample might appear contrived.  I can shed more light on what I was trying to accomplish if needed, but regardless, you still might want to look at this issue even if there is a better way to accomplish the task.

There are two projects in the sample, one utilizing EF7, the other EF6.  Both create an expression of 'OR' predicates and then executes a query using the expression in the where clause.  A test run is done for 1 through 26 predicates.  

The EF6 sample runs fine, however, on my machine, the EF7 sample starts to get very slow around 20 predicates.  Each additional predicate added seems to take approximately double the time of the previous run.

Here are the results for EF7:

TermCount 1 in 453 ms
TermCount 2 in 6 ms
TermCount 3 in 5 ms
TermCount 4 in 5 ms
TermCount 5 in 5 ms
TermCount 6 in 6 ms
TermCount 7 in 7 ms
TermCount 8 in 7 ms
TermCount 9 in 9 ms
TermCount 10 in 11 ms
TermCount 11 in 15 ms
TermCount 12 in 24 ms
TermCount 13 in 39 ms
TermCount 14 in 72 ms
TermCount 15 in 128 ms
TermCount 16 in 248 ms
TermCount 17 in 479 ms
TermCount 18 in 959 ms
TermCount 19 in 1903 ms
TermCount 20 in 3766 ms
TermCount 21 in 7562 ms
TermCount 22 in 15186 ms
TermCount 23 in 30131 ms
TermCount 24 in 60310 ms
TermCount 25 in 121548 ms
TermCount 26 in 245091 ms

Here is EF6:

TermCount 1 in 73 ms
TermCount 2 in 9 ms
TermCount 3 in 9 ms
TermCount 4 in 9 ms
TermCount 5 in 9 ms
TermCount 6 in 10 ms
TermCount 7 in 11 ms
TermCount 8 in 11 ms
TermCount 9 in 12 ms
TermCount 10 in 21 ms
TermCount 11 in 20 ms
TermCount 12 in 13 ms
TermCount 13 in 13 ms
TermCount 14 in 13 ms
TermCount 15 in 14 ms
TermCount 16 in 14 ms
TermCount 17 in 15 ms
TermCount 18 in 15 ms
TermCount 19 in 15 ms
TermCount 20 in 23 ms
TermCount 21 in 16 ms
TermCount 22 in 17 ms
TermCount 23 in 17 ms
TermCount 24 in 31 ms
TermCount 25 in 19 ms
TermCount 26 in 18 ms
   
### Further technical details

EF Core version: 1.0.0
Operating system: Windows 10
Visual Studio version: VS2015 Update 3

**Comment by ErikEJ** [comment]
Could you share the 2 sql statements?

**Comment by npnelson** [comment]
I can't say for sure, but I think the problem might be in how long it takes EF7 to calculate the SQL statement to execute, because when I look at SQL profiler, the actual query runs fast, it just takes awhile to appear in the profiler.

For 20 terms from SQL Profiler:

<b>EF7</b>
exec sp_executesql N'SELECT [param].[RowKey], [param].[Description]
FROM [TestModels] AS [param]
WHERE (((((((((((((((((((UPPER([param].[Description]) LIKE (N''%'' + @__itm_0) + N''%'' OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_1) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_2) + 
N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_3) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_4) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_5) + N''%'') OR 
UPPER([param].[Description]) LIKE (N''%'' + @__itm_6) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_7) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_8) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_9) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_10) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_11) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_12) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_13) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_14) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_15) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_16) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_17) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_18) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_19) + N''%'') OR UPPER([param].[Description]) LIKE (N''%'' + @__itm_20) + N''%''',N'@__itm_0 nvarchar(4000),@__itm_1 nvarchar(4000),@__itm_2 nvarchar(4000),@__itm_3 nvarchar(4000),@__itm_4 nvarchar(4000),@__itm_5 nvarchar(4000),@__itm_6 nvarchar(4000),@__itm_7 nvarchar(4000),@__itm_8 nvarchar(4000),@__itm_9 nvarchar(4000),@__itm_10 nvarchar(4000),@__itm_11 nvarchar(4000),@__itm_12 nvarchar(4000),@__itm_13 nvarchar(4000),@__itm_14 nvarchar(4000),@__itm_15 nvarchar(4000),@__itm_16 nvarchar(4000),@__itm_17 nvarchar(4000),@__itm_18 nvarchar(4000),@__itm_19 nvarchar(4000),@__itm_20 nvarchar(4000)',@__itm_0=N'T1',@__itm_1=N'T2',@__itm_2=N'T3',@__itm_3=N'T4',@__itm_4=N'T5',@__itm_5=N'T6',@__itm_6=N'T7',@__itm_7=N'T8',@__itm_8=N'T9',@__itm_9=N'T10',@__itm_10=N'T11',@__itm_11=N'T12',@__itm_12=N'T13',@__itm_13=N'T14',@__itm_14=N'T15',@__itm_15=N'T16',@__itm_16=N'T17',@__itm_17=N'T18',@__itm_18=N'T19',@__itm_19=N'T20',@__itm_20=N''

<b>EF6</b>
exec sp_executesql N'SELECT 
    [Extent1].[RowKey] AS [RowKey], 
    [Extent1].[Description] AS [Description]
    FROM [dbo].[TestModels] AS [Extent1]
    WHERE (UPPER([Extent1].[Description]) LIKE @p__linq__0 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__1 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__2 ESCAPE N''~'') OR 
(UPPER([Extent1].[Description]) LIKE @p__linq__3 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__4 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__5 ESCAPE N''~'') OR 
(UPPER([Extent1].[Description]) LIKE @p__linq__6 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__7 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__8 ESCAPE N''~'') OR 
(UPPER([Extent1].[Description]) LIKE @p__linq__9 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__10 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__11 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__12 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__13 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__14 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__15 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__16 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__17 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__18 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__19 ESCAPE N''~'') OR (UPPER([Extent1].[Description]) LIKE @p__linq__20 ESCAPE N''~'')',N'@p__linq__0 nvarchar(4000),@p__linq__1 nvarchar(4000),@p__linq__2 nvarchar(4000),@p__linq__3 nvarchar(4000),@p__linq__4 nvarchar(4000),@p__linq__5 nvarchar(4000),@p__linq__6 nvarchar(4000),@p__linq__7 nvarchar(4000),@p__linq__8 nvarchar(4000),@p__linq__9 nvarchar(4000),@p__linq__10 nvarchar(4000),@p__linq__11 nvarchar(4000),@p__linq__12 nvarchar(4000),@p__linq__13 nvarchar(4000),@p__linq__14 nvarchar(4000),@p__linq__15 nvarchar(4000),@p__linq__16 nvarchar(4000),@p__linq__17 nvarchar(4000),@p__linq__18 nvarchar(4000),@p__linq__19 nvarchar(4000),@p__linq__20 nvarchar(4000)',@p__linq__0=N'%T1%',@p__linq__1=N'%T2%',@p__linq__2=N'%T3%',@p__linq__3=N'%T4%',@p__linq__4=N'%T5%',@p__linq__5=N'%T6%',@p__linq__6=N'%T7%',@p__linq__7=N'%T8%',@p__linq__8=N'%T9%',@p__linq__9=N'%T10%',@p__linq__10=N'%T11%',@p__linq__11=N'%T12%',@p__linq__12=N'%T13%',@p__linq__13=N'%T14%',@p__linq__14=N'%T15%',@p__linq__15=N'%T16%',@p__linq__16=N'%T17%',@p__linq__17=N'%T18%',@p__linq__18=N'%T19%',@p__linq__19=N'%T20%',@p__linq__20=N'%%'

**Comment by maumar** [comment]
problem is indeed with the query compilation, namely EqualityPredicateInExpressionOptimizer and PredicateReductionExpressionOptimizer

**Comment by smitpatel** [comment]
For triage: This could be candidate for 1.0.1
It impact all the queries which has `OR` or `AND` in predicate. The execution time becomes proportional to n^2 in worst case.
The issue is, in `EqualityPredicateInExpressionOptimizer` & `PredicateReductionExpressionOptimizer`, we visit the subtrees but discard them if we cannot optimize and revisit them again in base visitor. 
Fix is to re-use the already visited children in making the returned expression. It is low risk fix and would return same tree. Just avoid evaluating same thing twice.
All tests passes successfully.
Following are perf numbers after the fix. (applied on 1.0.1 branch)
```
This test was run with SQL 2016 localDB.  Make sure you have it installed.  Alternatively, you could change the connection string in the code to point to a SQL database
Press <Enter> to run test

TermCount 1 in 476 ms
TermCount 2 in 6 ms
TermCount 3 in 6 ms
TermCount 4 in 5 ms
TermCount 5 in 5 ms
TermCount 6 in 6 ms
TermCount 7 in 6 ms
TermCount 8 in 7 ms
TermCount 9 in 7 ms
TermCount 10 in 8 ms
TermCount 11 in 13 ms
TermCount 12 in 9 ms
TermCount 13 in 9 ms
TermCount 14 in 10 ms
TermCount 15 in 10 ms
TermCount 16 in 10 ms
TermCount 17 in 11 ms
TermCount 18 in 11 ms
TermCount 19 in 11 ms
TermCount 20 in 12 ms
TermCount 21 in 17 ms
TermCount 22 in 13 ms
TermCount 23 in 15 ms
TermCount 24 in 13 ms
TermCount 25 in 14 ms
TermCount 26 in 14 ms
Test Complete
```

**Comment by smitpatel** [comment]
merged via #6256
